Add native-source package type

geoand opened the pull request:
This package type is meant to support workflows where the actual native-image is created by a different
container than the one that run that Quarkus build.

Resolves: #15208

geoand commented:
@jerboaa @ebaron here is the initial work for  #15208.

What it does create the same directory that the `native` package creates, but doesn't actually run the `native-image`.
Furthermore, it also writes a file named `native-image.args` inside that directory which contains all the native-image args (that don't actually depend on the version of the `native-image` binary).
You can then run the actual native-image run by doing something like:

```bash
$GRAALVM_HOME/bin/native-image $(cat native-image.args)
```

jerboaa commented on core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java:
```diff
@@ -81,6 +83,58 @@ ArtifactResultBuildItem result(NativeImageBuildItem image) {
+                .setNativeImageExecutable(Collections.emptyList())
+                // TODO: is this correct? It's set to true here because it's expected that native-sources will always be used
+                //  but it might make more sense to have some dedicated native-sources configuration
+                .setContainerBuild(true)
```
Any particular reason why `isContainerBuild` couldn't be used here instead of hard-coding `true`?

geoand replied:
I'll add it

zakkak commented:
LGTM, thanks @geoand 

When merging please keep in mind that this PR conflicts with https://github.com/quarkusio/quarkus/pull/14635 and https://github.com/quarkusio/quarkus/pull/13963 .

geoand commented:
I can wait for those to be merged if needed

geoand commented:
https://github.com/quarkusio/quarkus/pull/14635 was merged so I now need to rebase this onto the latest version of the code

geoand committed:
Add native-source package type

This package type is meant to support workflows where the actual native-image is created by a different
container than the one that run that Quarkus build.

Resolves: #15208

geoand commented:
PR has been rebased onto master and taken out of draft.

Please take a look :)

Thanks!

zakkak reviewed (approved):
LGTM, thanks @geoand

geoand commented:
Thanks for the review @zakkak

geoand commented:
@gsmet or @gastaldi mind taking a look at this as well?

Thanks

gastaldi reviewed (approved):
Nice! Loved it!

stuartwdouglas commented:
@geoand I missed this yesterday, but won't the native sources build not include the results of native image only build steps? If something is using `onlyIf = NativeBuild.class` they won't be included but should be?

geoand commented:
That's a good question, should they actually be included?

We would probably have to look at each one

geoand commented:
I'll go through them one by one and change the ones that need to be changed.

geoand commented:
Draft PR is here: https://github.com/quarkusio/quarkus/pull/15291

I want it to run on my fork before I request a review
